### ISSUE TITLE
Add Assert.NotEquals(default, <expr>) fixer for xUnit2002 diagnostics

### ISSUE THREAD

**Issue by jaredpar** [xunit/xunit#2010]
The `NotNull` fixer cannot be safely run in bulk because it will suggest removal of the entire statement. In the case the expression being targeted has side effects this could meaningfully alter the state of the tests. Consider:

```
public class UtilTest
{
    DateTime GetValue() => DateTime.UtcNow;

    [Fact]
    public void Example()
    {
        Assert.NotNull(GetValue());
    }
}
```

Here the xunit analyzer correctly flags the `Assert.NotNull` call as being non-sense. At the same time though fully removing the statement is questionable. If the `GetValue` method has side effects then that could potentially affect other code execution inside the method body. 

A safer alternative to consider is switching the call to `Assert.NotEqual`:

```
Assert.NotEqual(default, GetValue());
```

This will maintain the existing side affects and comparing a `struct` to `default` is the closest logical equivalent to comparing a reference type to `null`. 

This is the approach we took when cleaning up all of our xUnit2002 violations in Roslyn. There were ~2,000 of them and a good chunk of them had side effects in the expression. Many times it was the only expression of note in the test. Using `NotEqual(default)` helped us safely clean up the tests and in the process discover two bugs in our code base.

**Comment by bradwilson** [comment]
I don't think we're going to fix this.

My gut feeling here is that in the name of economy, you have combined two of the three As in AAA: Act and Assert. A test written like this in AAA style (comments are optional, just here to illustrate the style) would not have this issue:

```csharp
[Fact]
public void Example()
{
    // Arrange (nothing to do here)

    // Act
    var result = GetValue();

    // Assert
    Assert.NotNull(result);
}
```

**Comment by jaredpar** [comment]
2019 jaredpar is disappointed with this resolution 
2023 jaredpar is "meh" about the resolution. 

:)

In practice this hasn't been too much of a problem. When it has come up I've just disabled the suggestion and moved on to other things.

**Comment by bradwilson** [comment]
Out of curiosity, why not actually fix the issue (by hand) by removing the `Assert` while leaving the potentially side-effect'ing code? I'm not sure what the value of keeping the `Assert` is that you'd want to wrap it in a #pragma.

**Comment by jaredpar** [comment]
The motivation behind this issue is that roslyn changed on of our core types from a `class` to a `struct`. It was a critical perf improvement for the code base so had some urgency behind getting merged. This diagnostic revealed that there were several hundred unit tests that had the pattern `Assert.NotNull(MethodThatReturnedThatType())`. At first I went to the fix all to handle it but couldn't for the reasons stated in the issue. 

The diagnostic did call into question what exactly the tests were doing. The code paths clearly didn't have a `null` return (easy to verify when we flipped to `struct`). For the tests I dug into it was clear that really the important part was the side effect of the method executing and verifying that path didn't throw. The `Assert.NotNull` part was likely from a time in the past where it could return `null` on that path but that was likely an artifact of history. 

So definitely a case where the assert was questionable. At the same time there was pressure to get the change in, there were far too many cases to fix up by hand hence I believe we had to disable the diagnostic as a short term measure. The fix all I suggested here seemed like a valid alternative approach that would have allowed us to do a fix all hence filed it.

Note: this was 4-5 years ago so I'm sure I'm remembering some of the details wrong.
